core/src/com/google/inject/internal/InternalFactory.java

```diff
@@ -53,14 +60,127 @@ default Provider<T> makeProvider(InjectorImpl injector, Dependency<?> dependency
+   *     calls
+   * @param linked true if getting as a result of a linked binding
+   * @return a {@link MethodHandleResult} containing the method handle and its cachability settings
+   */
+  abstract MethodHandleResult makeHandle(LinkageContext context, boolean linked);
+
+  static MethodHandleResult makeCachable(MethodHandle methodHandle) {
+    return new MethodHandleResult(methodHandle, MethodHandleResult.Cachability.ALWAYS);
+  }
+  static MethodHandleResult makeCachableOnLinkedSetting(MethodHandle methodHandle) {
+    return new MethodHandleResult(methodHandle, MethodHandleResult.Cachability.ON_LINKED_SETTING);
+  }
+  static MethodHandleResult makeUncachable(MethodHandle methodHandle) {
+    return new MethodHandleResult(methodHandle, MethodHandleResult.Cachability.NEVER);
+  }
+  static final class MethodHandleResult {
+    static enum Cachability {
+      ALWAYS,
+      ON_LINKED_SETTING ,
+      NEVER,
+    }
+
+    
+    final MethodHandle methodHandle;
+    final Cachability cachability;
+
+    private MethodHandleResult(MethodHandle methodHandle, Cachability cachability) {
+      this.methodHandle = methodHandle;
+      this.cachability = cachability;
+    }
+  }
+
+
+  /** Racily read the cache. */
+  private MethodHandle readCache(boolean linked) {
+    var local = handleCacheResult;
+    if (local != null) {
+      if (local instanceof MethodHandle) {
+        return (MethodHandle) local;
+      }
+      var handleArray = (MethodHandle[]) local;
+      if (linked) {
+        return handleArray[1];
+      } else {
+        return handleArray[0];
+      }
+    }
+    return null;
+  }
+
+  /** Atomically update the cache and return the handle to use. */
+  private MethodHandle updateCache(boolean linked, MethodHandleResult result) {
+    switch (result.cachability) {
+      case NEVER:
+        return result.methodHandle;
+      case ALWAYS:
+        synchronized (this) {
+          var local = handleCacheResult;
+          if (local != null) {
+            if (local instanceof MethodHandle) {
+              // we lost a race
+              return (MethodHandle) local;
+            }
+            throw new AssertionError(
+                "Cannot update cache with an always cachable handle if it was already set with a"
+                    + " conditional cachable handle");
+          }
+          handleCacheResult = result.methodHandle;
+          return result.methodHandle;
+        }
+      case ON_LINKED_SETTING:
+        {
+          int index = linked ? 1 : 0;
+          synchronized (this) {
+            var local = handleCacheResult;
+            if (local != null) {
+              if (local instanceof MethodHandle[]) {
+                var array = (MethodHandle[]) local;
+                var handle = array[index];
+                if (handle != null) {
+                  // this implies we lost a race
+                  return handle;
+                }
+                array[index] = result.methodHandle;
```

Comment:
Is setting an item in an array a safe cross-thread operation (even if the array is volatile)?  I thought volatility was for the field itself, not the contents within it.  (Same below @ L177)